lib/idmapping.c

```diff
@@ -158,7 +158,7 @@ void write_mapping(int proc_dir_fd, int ranges, const struct map_range *mappings
 
 	/* Align setuid- and fscaps-based new{g,u}idmap behavior. */
 	if (geteuid() == 0 && geteuid() != ruid) {
-		if (prctl(PR_SET_KEEPCAPS, 1, 0, 0, 0) < 0) {
+		if (prctl(PR_SET_KEEPCAPS, 1L, 0L, 0L, 0L) == -1) {
```

Comment:
manpage says unsigned long :-)
https://man7.org/linux/man-pages/man2/prctl.2.html
```
       #include <sys/prctl.h>

       int prctl(int option, ...
                 /* unsigned long arg2, unsigned long arg3,
                 unsigned long arg4, unsigned long arg5 */ );
```

`L -> LU`

Reply:
I changed the manual page.

prctl(2) is actually a variadic function (the prototype that the user sees) in glibc:

```c
$ grepc prctl /usr/include/
/usr/include/x86_64-linux-gnu/sys/prctl.h:extern int prctl (int __option, ...) __THROW;
```

And the kernel later reinterprets the values as different types.  For example, `PR_GET_CHILD_SUBREAPER` reinterprets it as an `int *`:

```c
$ grep -rn -A2 PR_GET_CHILD_SUBREAPER kernel/sys.c 
2613:	case PR_GET_CHILD_SUBREAPER:
2614-		error = put_user(me->signal->is_child_subreaper,
2615-				 (int __user *)arg2);
```

So, the fact that it is stored as an `unsigned long` temporarily is irrelevant, as long as the width of the type is the same.

I split the prctl(2) page into one separate page for each operation in <https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/commit/?id=2534dc7ad5e9bed50006f5596385e640bd101b7d>.

And I chose to use the actual type with which the kernel will interpret the value, to document each operation:

```
$ MANWIDTH=72 man PR_GET_CHILD_SUBREAPER | cat
PR_GET_CHILD_SUBREAPER(2const)           PR_GET_CHILD_SUBREAPER(2const)

NAME
     PR_GET_CHILD_SUBREAPER  -  get  the "child subreaper" attribute of
     the calling process

LIBRARY
     Standard C library (libc, -lc)

SYNOPSIS
     #include <linux/prctl.h>  /* Definition of PR_* constants */
     #include <sys/prctl.h>

     int prctl(PR_GET_CHILD_SUBREAPER, int *isset);

DESCRIPTION
     Return the "child subreaper" setting of the caller, in  the  loca‐
     tion pointed to by isset.

RETURN VALUE
     On success, 0 is returned.  On error, -1 is returned, and errno is
     set to indicate the error.

ERRORS
     EFAULT
            isset is an invalid address.

STANDARDS
     Linux.

HISTORY
     Linux 3.4.

SEE ALSO
     prctl(2), PR_SET_CHILD_SUBREAPER(2const)

Linux man‐pages 6.8‐151‐g58... 2024‐06‐02PR_GET_CHILD_SUBREAPER(2const)
```

---

In some cases, the kernel uses a slightly inappropriate type, and I chose to document a compatible type that is more appropriate for user space.  In general, for bit masks of flags, I used `unsigned long` and for enumerations of booleans I used `long`.

In this case, PR_SET_KEEPCAPS receives a boolean value, so a signed integer is more appropriate.  User space doesn't need to deal with unsigned stuff here.

```
$ MANWIDTH=72 man PR_SET_KEEPCAPS | cat
PR_SET_KEEPCAPS(2const)                         PR_SET_KEEPCAPS(2const)

NAME
     PR_SET_KEEPCAPS - set the state of the "keep capabilities" flag

LIBRARY
     Standard C library (libc, -lc)

SYNOPSIS
     #include <linux/prctl.h>  /* Definition of PR_* constants */
     #include <sys/prctl.h>

     int prctl(PR_SET_KEEPCAPS, long state);

DESCRIPTION
     Set  the  state  of the calling thread’s "keep capabilities" flag.
     The effect of this flag is described  in  capabilities(7).   state
     must  be  either  0L  (clear  the flag) or 1L (set the flag).  The
     "keep capabilities" value will be reset to 0 on  subsequent  calls
     to execve(2).

RETURN VALUE
     On success, 0 is returned.  On error, -1 is returned, and errno is
     set to indicate the error.

ERRORS
     EINVAL
            state is not a valid value.

     EPERM  The caller’s SECBIT_KEEP_CAPS_LOCKED flag is set (see capa‐
            bilities(7)).

STANDARDS
     Linux.

HISTORY
     Linux 2.2.18.

SEE ALSO
     prctl(2), PR_GET_KEEPCAPS(2const)

Linux man‐pages 6.8‐151‐g58... 2024‐06‐02       PR_SET_KEEPCAPS(2const)
```

Reply:
> I changed the manual page. :-)

That is also  way to keep documentation and real usage the same :-)
I like consistency 

+1